File: server/src/main/java/ru/fullrest/mfr/server/service/PropertyService.java
package ru.fullrest.mfr.server.service;

import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import ru.fullrest.mfr.server.configuration.CacheConfiguration;
import ru.fullrest.mfr.server.model.entity.Property;
import ru.fullrest.mfr.server.model.entity.PropertyType;
import ru.fullrest.mfr.server.model.repository.PropertyRepository;

import javax.transaction.Transactional;

@Service
@RequiredArgsConstructor
public class PropertyService {

    private final PropertyRepository propertyRepository;

    @Cacheable(value = CacheConfiguration.PROPERTY_CACHE, unless = "#result == null")
    public Property findByType(PropertyType type) {
        return propertyRepository.findByType(type).orElse(null);
    }

    @Transactional
    @CacheEvict(value = CacheConfiguration.PROPERTY_CACHE, allEntries = true)
    public void save(Property property) {
        propertyRepository.save(property);
    }
}

File: server/src/main/kotlin/com/lezenford/mfr/server/service/model/PropertyService.kt
package com.lezenford.mfr.server.service.model

import com.lezenford.mfr.server.configuration.CacheConfiguration
import com.lezenford.mfr.server.model.entity.Property
import com.lezenford.mfr.server.model.repository.PropertyRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class PropertyService(
    private val propertyRepository: PropertyRepository
) {
    @Cacheable(value = [CacheConfiguration.PROPERTY_CACHE], unless = "#result == null")
    suspend fun findByType(type: Property.Type): Property? {
        return propertyRepository.findByType(type)
    }

    @Transactional
    @CacheEvict(value = [CacheConfiguration.PROPERTY_CACHE], allEntries = true)
    suspend fun save(property: Property) {
        propertyRepository.save(property)
    }
}